milli/src/vector/parsed_vectors.rs

```diff
@@ -0,0 +1,208 @@
+            return Err(Error::InvalidMap(value));
+        };
+        Ok(ParsedVectors(value))
+    }
+
+    pub fn retain_user_provided_vectors(&mut self) {
+        self.0.retain(|_k, v| match v {
+            Vectors::ImplicitlyUserProvided(_) => true,
+            Vectors::Explicit(ExplicitVectors { embeddings: _, user_provided }) => *user_provided,
+        });
+    }
+}
+
+pub enum Error {
+    InvalidMap(Value),
+    InternalSerdeJson(serde_json::Error),
+}
+
+impl Error {
+    pub fn to_crate_error(self, document_id: String) -> crate::Error {
+        match self {
+            Error::InvalidMap(value) => {
+                crate::Error::UserError(UserError::InvalidVectorsMapType { document_id, value })
+            }
+            Error::InternalSerdeJson(error) => {
+                crate::Error::InternalError(InternalError::SerdeJson(error))
+            }
+        }
+    }
+}
+
+fn to_vector_map(
+    obkv: KvReaderDelAdd,
+    side: DelAdd,
+) -> Result<Option<BTreeMap<String, Vectors>>, Error> {
+    Ok(if let Some(value) = obkv.get(side) {
+        let ParsedVectors(parsed_vectors) = ParsedVectors::from_bytes(value)?;
+        Some(parsed_vectors)
+    } else {
+        None
+    })
+}
+
+/// Represents either a vector or an array of multiple vectors.
+#[derive(serde::Serialize, serde::Deserialize, Debug)]
+#[serde(transparent)]
+pub struct VectorOrArrayOfVectors {
+    #[serde(with = "either::serde_untagged_optional")]
+    inner: Option<either::Either<Embedding, Vec<Embedding>>>,
+}
+
+impl VectorOrArrayOfVectors {
+    pub fn into_array_of_vectors(self) -> Option<Vec<Embedding>> {
+        match self.inner? {
+            either::Either::Left(vector) => Some(vec![vector]),
+            either::Either::Right(vectors) => Some(vectors),
+        }
+    }
+
+    pub fn from_array_of_vectors(array_of_vec: Vec<Embedding>) -> Self {
+        Self { inner: Some(either::Either::Right(array_of_vec)) }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::VectorOrArrayOfVectors;
+
+    #[test]
+    fn array_of_vectors() {
+        let null: VectorOrArrayOfVectors = serde_json::from_str("null").unwrap();
+        let empty: VectorOrArrayOfVectors = serde_json::from_str("[]").unwrap();
+        let one: VectorOrArrayOfVectors = serde_json::from_str("[0.1]").unwrap();
+        let two: VectorOrArrayOfVectors = serde_json::from_str("[0.1, 0.2]").unwrap();
+        let one_vec: VectorOrArrayOfVectors = serde_json::from_str("[[0.1, 0.2]]").unwrap();
+        let two_vecs: VectorOrArrayOfVectors =
+            serde_json::from_str("[[0.1, 0.2], [0.3, 0.4]]").unwrap();
+
+        insta::assert_json_snapshot!(null.into_array_of_vectors(), @"null");
+        // 👇 is the the intended behavior? would rather expect [] here, but changing that is a breaking change...
+        insta::assert_json_snapshot!(empty.into_array_of_vectors(), @r###"
+        [
+          []
+        ]
+        "###);
```

Comment:
Can we change this to return a single `[]`, please?

Reply:
done